### PR TITLE
Modern RabbitMQ

### PR DESCRIPTION
**Pull request by binarin**
###### Motivation for this change

This series of commits upgrades RabbitMQ to latest stable version and modernizes corresponding NixOS module. Most important changes are:
- Proper integration with systemd: socket-activated epmd, startup notifications, sane defaults
- Using new configuration file format, which is more automation friendly

Parts of Erlang infrastructure were also touched to make socket-activation real.

Every commit message also contains more detailed description, together with relevant links to RabbitMQ and Erlang/OTP repositories.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
@GrahamcOfBorg build rabbitmq

**Comment by grahamc** [comment]
...
@GrahamcOfBorg build rabbitmq_server

**Comment by adisbladis** [comment]
Tests are passing on `x86_64-linux` 👍 

I'm very much in favour of this PR but I want @grahamc to give the final OK. :)

**Comment by binarin** [comment]
Bumped rabbit to recently released 3.7.3, moved elixir version selection to all-packages.nix as suggested.

**Comment by Profpatsch** [comment]
Since `3.6` [is going to be supported until May 2018](https://groups.google.com/forum/#!msg/rabbitmq-users/kXkI-f3pgEw/UFowJIK4BQAJ), I request that an up-to-date `3.6.x` is kept in nixpkgs.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
Since `3.6` [is going to be supported until May 2018](https://groups.google.com/forum/#!msg/rabbitmq-users/kXkI-f3pgEw/UFowJIK4BQAJ), I request that an up-to-date `3.6.x` is kept in nixpkgs.

Here is a patch to 3.6.14: https://gitlab.techcultivation.org/Profpatsch/nixpkgs-techcultivation/commit/38afdc9923f3b6aa601023b599a34dcc4af58f96

**Comment by binarin** [comment]
@Profpatsch I can easily do this for package itself, but I don't think I can do this in a clean way for NixOS module. If you are OK with that, how'd you prefer the new and the old package named in all-packages?

**Comment by Profpatsch** [comment]
> but I don't think I can do this in a clean way for NixOS module.

I’d to it like this:

For the packages:
```
let
  majorVersion = 6;
in … {
  version = "${toString majorVersion}.x.x";
  …
  passthrough = { inherit majorVersion; };
}
```

And then use `cfg.package.majorVersion >= 7` to switch in the module.

> how'd you prefer the new and the old package named in all-packages?

They should probably be named
```
rabbitmq-server = rabbitmq-server_3_6;
rabbitmq-server_3_6 = …
rabbitmq-server_3_7 = …
```

Notice that I changed the (historical) `_` to `-`, that needs to be documented in the release notes under `nixos/doc/manual/release-notes/`.

**Comment by Profpatsch** [comment]
fyi: I’ve created a PR to update to `3.6.15`, with various improvements to the package. Some of them are probably valid for `3.7` as well and should be merged into the `3.7` derivation as well.

I intend to merge the `3.6` update first, since we should support the latest release of that major release (see above).

**Comment by peterhoeg** [comment]
While we're at it, can we use ```erlang_nox``` instead of regular ```erlang``` by default when building rabbitmq? We've been using that for the last 6 months or so without any problems.

**Comment by Profpatsch** [comment]
I’m for it, I [already disable wx for our rabbitmq docker image](https://gitlab.techcultivation.org/sangha/sangha-deployment/blob/bd04f830d0bb46e4152a8710fcae7aa26e163f61/containers/rabbitmq.nix#L26).

As you can see in that source file, the wx beam files are put in nonetheless, so `erlang_nox` should probably also throw them out if possible. Rabbitmq itself depends on `coreutils` because of one command used in a wrapper script, that can probably be patched out as well.

**Comment by Profpatsch** [comment]
The deprecation for the `3.6` series is coming nearer (May 1st), so we should get this into `nixpkgs`.

We should still keep `3.6` around (though with a deprecation warning) and remove it with the next release (`18.09`).

File: nixos/modules/services/networking/epmd.nix [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+    systemd.services.epmd = {
+      description = "Erlang Port Mapper Daemon";
+      after = [ "network.target" ];
+      requires = [ "epmd.socket" ];
+
+      serviceConfig = {
+        User = "nobody";
+        Group = "nogroup";
```
Comment (peterhoeg) [inline nixos/modules/services/networking/epmd.nix]: We shouldn't be using nobody/nogroup - those are reserved for NFS. Instead either run with ```DynamicUser = true;``` (preferred) if possible or alternatively create a user/group for this.

File: nixos/modules/services/networking/epmd.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        ExecStart = "${cfg.package}/bin/epmd -systemd";
+        Type = "notify";
+        StandardOutput = "journal";
+        StandardError = "journal";
```
Comment (peterhoeg) [inline nixos/modules/services/networking/epmd.nix]: ```journal``` is default anyway, so you don't have to set those.

**Comment by binarin** [comment]
I updated this patch without re-introducing version-selection logic described above, it's already to late.

This depends on:
- https://github.com/NixOS/nixpkgs/pull/43817
- https://github.com/NixOS/nixpkgs/pull/43812

`ps` with `systemd` support is now required because of https://github.com/rabbitmq/rabbitmq-server/blob/7c941aff893c44355e3e892a43b80d20de171ac3/src/rabbit.erl#L415

**Comment by binarin** [comment]
@Profpatsch All dependencies are in `master` now, and I've rebased/updated the PR itself.

**Review by Profpatsch — APPROVED** [review]

**Comment by Profpatsch** [comment]
@GrahamcOfBorg build rabbitmq-server

**Comment by Profpatsch** [comment]
I think the darwin failure is unrelated. Let’s try again.

@GrahamcOfBorg build rabbitmq-server

**Pull request merged**